### ISSUE TITLE
zfs-dracut on CentOS 6.5 broken but I have a few fixes

### ISSUE THREAD

**Issue by srlefevre** [zfsonlinux/zfs#2356]
As reported in issue #2312 , RHEL/CentOS 6.x uses /usr/share/dracut/modules.d and not /lib/dracut/modules.d.  

RHEL/CentOS appears to be using a an older version of dracut (v.004) relative to other distros which is likely why zfs-dracut is broken on RHEL/CentOS.

I've done a fair amount of testing/fixing to get CentOS to use a zfs root file system with success using zfs-0.6.2-1.el6.x86_64 and related rpms.  I'm booting from an ext4 partition but then import a zpool and mounting a zfs fs as root with other zfs fs auto mounting (e.g. /home /opt).

I'd like to get my few fixes melded into the zfs-dracut for the long term. 

What is the best way to contribute?

Issues I've found and possibly addressed:
1. dracut/modules.d/90zfs should be under /usr/share/
Fix: ln -s /lib/dracut/modules.d/90zfs /usr/share/dracut/modules.d/90zfs
(Could be addressed in the .spec file to make the rpm relocatable?)

2. /bin/awk missing from initramfs so module spl doesn't load via dracut but will load once zfs init script is run.
Fix: 
```
$ diff /lib/dracut/modules.d/90zfs/module-setup.sh /usr/share/dracut/modules.d/90zfs/module-setup.sh 
39a40,41
>       dracut_install /bin/gawk
>       dracut_install /bin/awk
(awk is a symlink'd to gawk)
```

3. The older dracut uses separate scripts (check, install, installkernel) instead of module-setup.sh
Fix: create stub scripts in 90zfs for each along the line of:
```
$ cat install
#!/bin/bash
. /usr/share/dracut/modules.d/90zfs/module-setup.sh
install $1
```
4. mount-zfs.sh uses the function getargbool but its not defined in the older dracut so call fails and the 'zfs_force' kernel cmdline arg does not work.
Fix: none yet. Hack is to set ZPOOL_FORCE="-f" at top of script.

5. parse-zfs.sh uses /initqueue/finished/ but the older dracut doesn't have that directory and uses /initqueue-finished/ instead.
Fix: 
```
$ diff /lib/dracut/modules.d/90zfs/parse-zfs.sh /usr/share/dracut/modules.d/90zfs/parse-zfs.sh 
57c57
<       echo '[ -e /dev/zfs ]' > $hookdir/initqueue/finished/zfs.sh
---
>       echo '[ -e /dev/zfs ]' > $hookdir/initqueue-finished/zfs.sh
```

I don't know if some of these issues can be addressed in the build process and make them conditional depending on the target or possibly in the .spec files based on dependencies/version of dracut during install.  Or yet still possibly conditionally during the execution of dracut via the functions in module-setup.sh.

**Comment by Rudd-O** [comment]
Do we still support CentOS 6?

If we do, mind producing properly formatted patches?  Thanks.  I can't read them because of the formatting.  Use triple backquotes on a single line to begin and end your code blocks to get them to format right, or submit a pull request instead.

**Comment by behlendorf** [comment]
@Rudd-O yes we do.  We'll likely continue to support it until it goes end of life in 2020 or we're forced to drop support for technical reasons such as upstream kernel compatibility.

I've edited the above comment to add the needed formatting.

**Comment by Rudd-O** [comment]
The dracut directory being /usr/share is a matter that is resolved in the specfile zfs.spec:

```
%if 0%{?_dracut}
# Set the default dracut directory based on distribution.
%if %{undefined _dracutdir}
%if 0%{?fedora} >= 17 || 0%{?rhel} >= 7 || 0%{?centos} >= 7
%global _dracutopt  --with-dracutdir=%{_prefix}/lib/dracut
%else
%global _dracutopt  --with-dracutdir=%{_prefix}/share/dracut
%endif
%else
%global _dracutopt  --with-dracutdir=%{_dracutdir}
%endif
%else
%global _dracutopt ''
%endif
```

**Comment by Rudd-O** [comment]
awk is not needed by any Dracut script as it currently stands in the `topic-fixboot` branch (pull request) or systemd unit file.  This should not be a problem.

**Comment by Rudd-O** [comment]
I have pushed a fix to the initqueue issue into the `topic-fixboot` pull request.

**Comment by Rudd-O** [comment]
These are the issues that remain to be addressed:

1. The older dracut uses separate scripts (check, install, installkernel) instead of module-setup.sh.
2. `mount-zfs.sh` uses the function `getargbool` but its not defined in the older dracut so call fails and the `zfs_force` kernel cmdline arg does not work.

Fix for #1 — fix: create stub scripts in 90zfs for each along the line of:

```
$ cat install
#!/bin/bash
. /usr/share/dracut/modules.d/90zfs/module-setup.sh
install $1
```

Fix for #2: none yet. Hack is to set `ZPOOL_FORCE="-f"` at top of script.

With that, we will have fully functional Red Hat  6.0 support for booting off of ZFS root.

**Comment by Rudd-O** [comment]
I experimented with creating the stubs, and I am sorry to say, I cannot fix that bug because my autotools knowledge cannot cross this hurdle:

Any script named `install`, `check` or `depends` causes the generated Makefile to throw `make` into an infinite loop.

Stands to reason, if you think about it.  When `install` becomes a target, then obviously it's no longer the `.PHONY` target that causes install to happen, so calling `make all` implicitly calls `make install`, which goes into a recursive loop.

Here is the diff that causes the issue:

```
diff --git a/contrib/dracut/90zfs/Makefile.am b/contrib/dracut/90zfs/Makefile.am
index f81d6c3..15f4bad 100644
--- a/contrib/dracut/90zfs/Makefile.am
+++ b/contrib/dracut/90zfs/Makefile.am
@@ -1,5 +1,9 @@
 pkgdracutdir = $(dracutdir)/modules.d/90zfs
 pkgdracut_SCRIPTS = \
+	check \
+	depends \
+	installkernel \
+	install \
 	export-zfs.sh \
 	module-setup.sh \
 	mount-zfs.sh \
@@ -8,6 +12,10 @@ pkgdracut_SCRIPTS = \
 	zfs-lib.sh
 
 EXTRA_DIST = \
+	$(top_srcdir)/contrib/dracut/90zfs/check.in \
+	$(top_srcdir)/contrib/dracut/90zfs/depends.in \
+	$(top_srcdir)/contrib/dracut/90zfs/installkernel.in \
+	$(top_srcdir)/contrib/dracut/90zfs/install.in \
 	$(top_srcdir)/contrib/dracut/90zfs/export-zfs.sh.in \
 	$(top_srcdir)/contrib/dracut/90zfs/module-setup.sh.in \
 	$(top_srcdir)/contrib/dracut/90zfs/mount-zfs.sh.in \
diff --git a/contrib/dracut/90zfs/check.in b/contrib/dracut/90zfs/check.in
index e69de29..0c996a3 100755
--- a/contrib/dracut/90zfs/check.in
+++ b/contrib/dracut/90zfs/check.in
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+NAME="$( basename "${BASH_SOURCE[0]}" )"
+
+. "$DIR/module-setup.sh"
+"$NAME" "$@"
diff --git a/contrib/dracut/90zfs/depends.in b/contrib/dracut/90zfs/depends.in
index e69de29..0c996a3 100755
--- a/contrib/dracut/90zfs/depends.in
+++ b/contrib/dracut/90zfs/depends.in
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+NAME="$( basename "${BASH_SOURCE[0]}" )"
+
+. "$DIR/module-setup.sh"
+"$NAME" "$@"
diff --git a/contrib/dracut/90zfs/install.in b/contrib/dracut/90zfs/install.in
index e69de29..0c996a3 100755
--- a/contrib/dracut/90zfs/install.in
+++ b/contrib/dracut/90zfs/install.in
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+NAME="$( basename "${BASH_SOURCE[0]}" )"
+
+. "$DIR/module-setup.sh"
+"$NAME" "$@"
diff --git a/contrib/dracut/90zfs/installkernel.in b/contrib/dracut/90zfs/installkernel.in
index e69de29..0c996a3 100755
--- a/contrib/dracut/90zfs/installkernel.in
+++ b/contrib/dracut/90zfs/installkernel.in
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+NAME="$( basename "${BASH_SOURCE[0]}" )"
+
+. "$DIR/module-setup.sh"
+"$NAME" "$@"
```

I will need more help from someone interested in Red Hat 6.0 support, as that's as far as I know how to go.

**Comment by Rudd-O** [comment]
A compatibility `getargbool` has been pushed to branch `topic-fixboot`.

The last remaining issue is the build system issue.  Need help with that.  Sorry.

**Comment by behlendorf** [comment]
Most of these issues have been addressed by d402c18.  I've opened #4640 to track the remaining issues.
